File: ios/LivingRoomEdge/LivingRoomEdge/Skill/Skill.swift
import Foundation

struct SkillContext {
    let edgeId: String
    let planId: String
    let stepId: String
}

struct SkillResult {
    let ok: Bool
    let message: String?
    let outputs: [String: String]?

    static func ok(_ message: String? = nil, outputs: [String: String]? = nil) -> SkillResult {
        SkillResult(ok: true, message: message, outputs: outputs)
    }

    static func error(_ message: String) -> SkillResult {
        SkillResult(ok: false, message: message, outputs: nil)
    }
}

protocol Skill: AnyObject {
    /// Wire contract: Service → Capability (schemas).
    func service() -> ServiceDescriptor
    /// Invoke by capability_id (local debug may also pass legacy action names).
    func execute(capabilityId: String, params: [String: String], context: SkillContext) async -> SkillResult
}

File: android/app-v2/src/main/java/com/smarthome/livingroom_v2/skill/Skill.kt
package com.smarthome.livingroom_v2.skill

import android.content.Context
import com.smarthome.livingroom_v2.brain.dto.ServiceDescriptor

data class SkillContext(
    val appContext: Context,
    val edgeId: String,
    val planId: String,
    val stepId: String,
)

data class SkillResult(
    val ok: Boolean,
    val message: String? = null,
    val outputs: Map<String, String> = emptyMap(),
) {
    companion object {
        fun ok(message: String? = null, outputs: Map<String, String> = emptyMap()) =
            SkillResult(ok = true, message = message, outputs = outputs)

        fun error(message: String) = SkillResult(ok = false, message = message)
    }
}

/**
 * Local executable service plugin.
 * Wire contract is [service]; invoke with [capabilityId] (not legacy action names).
 *
 * Runtime MUST call [isAvailable] before [execute]. Default: available.
 */
interface Skill {
    fun service(): ServiceDescriptor

    /**
     * Fast preflight. Default returns ok.
     * Override for hardware that can be probed cheaply (e.g. camera AP).
     */
    suspend fun isAvailable(
        capabilityId: String,
        params: Map<String, Any?>,
        ctx: SkillContext,
    ): SkillResult = SkillResult.ok("available")

    suspend fun execute(
        capabilityId: String,
        params: Map<String, Any?>,
        ctx: SkillContext,
    ): SkillResult
}
